---
title: PBMC QC, clustering and filtering
output: 
  md_document
---

```{r knitr opts, echo = FALSE, warning = FALSE, message = FALSE}
library("here")
knitr::opts_chunk$set(
  fig.width = 12,
  fig.height = 9,
  fig.path = paste0(fs::path_rel(here("scripts/data-scripts/rmd_files/pbmc/")), "/")
)
source(here(".Rprofile"), chdir = TRUE)
```

```{r libs}

library("edgeR")
library("scran")
library("scater")
library("Seurat")
library("viridis")
library("here")
library("dplyr")

# https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.1.0/5k_pbmc_protein_v3

options(stringAsFactors = FALSE)
```


```{r data}
assays <- Read10X(data.dir = here("downloads/5k_pbmc_protein_v3_filtered_feature_bc_matrix/"))
pbmc <- CreateSeuratObject(assays[["Gene Expression"]],
  project = "pbmc", 
  min.cells = 3,
  min.features = 200
)

```

```{r qc}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))
```

```{r normalise}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & percent.mt < 25)
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

```{r pca}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
DimPlot(pbmc, reduction = "pca")
DimHeatmap(pbmc, dims = 1:28, cells = 500, balanced = TRUE)
pbmc <- JackStraw(pbmc, num.replicate = 100, dims = 50)
pbmc <- ScoreJackStraw(pbmc, dims = 1:50)
JackStrawPlot(pbmc, dims = 1:28)
ElbowPlot(pbmc, ndims = 50)
```

```{r cluster}
pbmc <- FindNeighbors(pbmc, dims = 1:20)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

```{r umap}
pbmc <- RunUMAP(pbmc, dims = 1:20)


DimPlot(pbmc, reduction = "umap", cols = "Paired")
```

```{r markers}
pbmc_markers <- FindAllMarkers(
  pbmc,
  only.pos = TRUE,
  min.pct = 0.25,
  logfc.threshold = 0.25
)

g3 <- pbmc_markers %>% 
  filter(cluster == 0) %>% 
  top_n(n = 3, wt = avg_logFC) %>%
  pull(gene)

g5 <- pbmc_markers %>% 
  filter(cluster == 0) %>% 
  top_n(n = 5, wt = avg_logFC) %>%
  pull(gene)

VlnPlot(pbmc, features = g3, slot = "counts", log = TRUE)
FeaturePlot(pbmc, features = g5) + scale_color_viridis()

top10 <- pbmc_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()

t_cells <- pbmc[, FetchData(pbmc, "ident") == 0]


counts <- t_cells@assays$RNA@counts

```



```{r reload_data}

x10 <- read10X(
  path = here("downloads/5k_pbmc_protein_v3_filtered_feature_bc_matrix/")
)

sce <- SingleCellExperiment(
  assays = list(counts = x10$counts),
  colData = x10$samples,
  rowData = x10$genes
)
colnames(sce) <- gsub("-1", "", sce$Barcode)

sce <- sce[, colnames(counts)]
```

```{r filter}
counts(sce) <- as.matrix(counts(sce))
libsize_drop <- isOutlier(
  Matrix::colSums(counts(sce)),
  nmads = 3,
  type = "lower",
  log = TRUE
)
feature_drop <- isOutlier(
  Matrix::colSums(counts(sce) != 0),
  nmads = 3,
  type = "lower",
  log = TRUE
)
ind_expressed <- rowMeans(counts(sce) != 0) > 0.2 & 
  Matrix::rowMeans(counts(sce)) > 0.1
ind_drop <- libsize_drop | feature_drop

sce <- sce[ind_expressed, !ind_drop]
```


```{r save}
pbmc <- sce
saveRDS(pbmc, here("data/pbmc.rds"))
# usethis::use_data(pbmc, overwrite = TRUE)
# saveRDS(sce, "datasets/pbmc_nk_data.rds")

```
